### PR TITLE
Add test for detecting inconsistencies between declarative and programmatic config

### LINKED ISSUE #18221: Add test for detecting inconsistencies between declarative and programmatic config
Starting with 4.2, declarative configuration reuses the `XXXConfig` instances set on their parent config instance, in order to support config overrides. If these _initial instances_ change any attribute compared to the default `XXXConfig` instance, there is potentially a behavioral inconsistency between programmatic and declarative configuration. An example of this is #18216.

To ensure there is no regression, we need to write a test or tests that check for these possible inconsistencies. A reflection-based test would be the best - if possible - that requires minimal or no updates when introducing a new entry. The idea is to check like this:
```
// compare default and initial
assertEquals(new MapStoreConfig(), new MapConfig().getMapStoreConfig());
```
This should have caught the referenced #18216.

### PR DESCRIPTION
**Pull request by ramizdundar**
Tests if `new XXXConfig()` (default) different from `parentConfig.getXXXConfig()` (initial) with reflection.

Fixes #18221

### CODE REVIEW

**Comment by ramizdundar** [comment]
There are 4 kind of differences:

**Case 1**: There is no same argument constructor with get method

- com.hazelcast.config.SplitBrainProtectionConfig (Child of com.hazelcast.config.Config) does not have a suitable constructor. Note that getter of this config doesn't return null.
- com.hazelcast.config.security.CredentialsIdentityConfig (Child of com.hazelcast.client.config.ClientSecurityConfig) does not have a suitable constructor.
- com.hazelcast.config.security.TokenIdentityConfig (Child of com.hazelcast.client.config.ClientSecurityConfig) does not have a suitable constructor.
- com.hazelcast.config.security.UsernamePasswordIdentityConfig (Child of com.hazelcast.client.config.ClientSecurityConfig) does not have a suitable constructor.


**Case 2**: There are abstract config classes

- Error occurred while calling the constructor of com.hazelcast.config.SecureStoreConfig (Child of com.hazelcast.config.EncryptionAtRestConfig)

**Case 3**: There are configurations that take multiple parameters inside getter methods. (Special case of case 1)

- com.hazelcast.config.QueryCacheConfig (Child of com.hazelcast.client.config.ClientConfig) called from getOrNullQueryCacheConfig
- com.hazelcast.config.QueryCacheConfig (Child of com.hazelcast.client.config.ClientConfig) called from getOrCreateQueryCacheConfig

**Case 4**: There are situations where default and initial configs are not the same. These are the cases we initially want to test.

- com.hazelcast.config.NetworkConfig (Child of com.hazelcast.config.Config) failed the test.
- com.hazelcast.config.AdvancedNetworkConfig (Child of com.hazelcast.config.Config) failed the test.
- com.hazelcast.config.EvictionConfig (Child of com.hazelcast.config.MapConfig) failed the test.
- com.hazelcast.config.MapStoreConfig (Child of com.hazelcast.config.MapConfig) failed the test.
- com.hazelcast.config.RingbufferStoreConfig (Child of com.hazelcast.config.RingbufferConfig) failed the test.
- com.hazelcast.config.MergePolicyConfig (Child of com.hazelcast.config.CardinalityEstimatorConfig) failed the test.

**Comment by ramizdundar** [comment]
There are 5 situations where default and initial configs differ:

- com.hazelcast.config.SplitBrainProtectionConfig (Child of com.hazelcast.config.Config) failed the test. Because the constructor we added to `SplitBrainProtectionConfig` has `this.enabled = true`. Similar to `MapStoreConfig`.
- com.hazelcast.config.EvictionConfig (Child of com.hazelcast.config.MapConfig) failed the test. Because size is set to `UNSET` in `EvictionConfig`(default) and `DEFAULT_MAX_SIZE` in `MapConfig`(initial).
- com.hazelcast.config.MapStoreConfig (Child of com.hazelcast.config.MapConfig) failed the test. #18216 
- com.hazelcast.config.RingbufferStoreConfig (Child of com.hazelcast.config.RingbufferConfig) failed the test. Similar to `MapStoreConfig`.
- com.hazelcast.config.MergePolicyConfig (Child of com.hazelcast.config.CardinalityEstimatorConfig) failed the test. Because policies differ. `PutIfAbsentMergePolicy` in `MergePolicyConfig`(default) and `HyperLogLogMergePolicy` in `CardinalityEstimatorConfig`(initial)

File: hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java:142]: Can be simplified to `throws Exception`.

File: hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java:50]: This one shouldn't be `static`.

File: hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/config/DefaultConfigValidationTest.java:57]: I think we're better to add a short comment for each why they are ignored before we forget the reason.

**Review by blazember — CHANGES_REQUESTED** [review]
LGTM, added a few minors and one on `SplitBrainProtectionConfig` that I think is a non-valid change.

File: hazelcast/src/main/java/com/hazelcast/config/SplitBrainProtectionConfig.java
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/config/SplitBrainProtectionConfig.java:87]: I thought just like how `MapStoreConfig` has enabled by default this should also enabled but I will change.

**Review by Fly-Style — APPROVED** [review]
LGTM except notes from @blazember.
Good job, thanks!

**Comment by mmedenjak** [comment]
@ramizdundar moving to the next milestone to clear up the non-critical issues and PRs. But please keep working on it and get it reviewed. If it gets approved before the release branch is created, just merge it to master and change the milestone back for the PR and issue.

**Comment by ramizdundar** [comment]
@mmedenjak ok hopefully it'll get approved soon if nothing goes wrong.

**Review by blazember — APPROVED** [review]
LGTM, thanks @ramizdundar for picking this up :+1:

**Pull request merged**
